core/query/query.ml

```diff
@@ -865,27 +885,43 @@ struct
         | (Constant.Float a , Constant.Float b)  -> bool (a = b)
         | (Constant.Char a  , Constant.Char b)   -> bool (a = b)
         | (Constant.String a, Constant.String b) -> bool (a = b)
-        | (a, b)                 -> Apply ("==", [Constant a; Constant b])
+        | (a, b)                 -> Apply (Primitive "==", [Constant a; Constant b])
     in
-      match a, b with
+      match norm env a, norm env b with
         | (Constant a, Constant b) -> eq_constant (a, b)
         | (Variant (s1, a), Variant (s2, b)) ->
           if s1 <> s2 then
             Constant (Constant.Bool false)
           else
-            reduce_eq (a, b)
+            reduce_eq env (a, b)
         | (Record lfields, Record rfields) ->
           List.fold_right2
             (fun (_, v1) (_, v2) e ->
-              reduce_and (reduce_eq (v1, v2), e))
+              reduce_and env (reduce_eq env (v1, v2), e))
             (StringMap.to_alist lfields)
             (StringMap.to_alist rfields)
             (Constant (Constant.Bool true))
-        | (a, b) -> Apply ("==", [a; b])
+        | (a, b) -> Apply (Primitive "==", [a; b])
+  and norm_comp env c = norm env (computation env c)
+
+(*
+  let norm env t0 =
+    begin
+      Printf.printf "NORM: input %s\n" (string_of_t t0);
+      let res = norm env t0 in
+      Printf.printf "NORM: output %s\n\n" (string_of_t res);
+      res
+    end
+	*)
 
   let eval env e =
 (*    Debug.print ("e: "^Ir.show_computation e); *)
-    computation (env_of_value_env env) e
+    norm_comp (env_of_value_env env) e
+
+  let reduce_where_then = reduce_where_then (Value.Env.empty, Env.Int.empty)
```

Comment:
I'd avoid using the same name for the function result and function name, even though in this bounded / nonrecursive let it is technically fine.

Reply:
This is a case of encapsulation + shadowing of a recursive function, in which we want to prevent some technical details from being visible. In these cases I often use the same name for the implementation and the encapsulating function (rather than e.g. finding an alternate name for the recursive implementation which would be likely to be a bit silly). 
It is certainly a practice I don't want to abuse, but I believe it should be ok here?

Reply:
Oh, I see, this is  due to `reduce_where_then` being exported, and you are rebinding them to sensible-to-call-remotely partially applied versions.  If my earlier suggestion (about decoupling `reduce_*` from normalization) works out then this would go away.  I guess this assumes that the places where they are called elsewhere are also normalizing the argument(s) before calling (or that normalization is unnecessary).  However, I think before the `reduce_` functions were not recursively calling translation/normalization anyway so perhaps this is further reason to revert and transplant them to `Query.Lang`.